### DISCUSSION TITLE
Getting type error when using generics with mutationFn

### DISCUSSION THREAD

**Discussion by ThreeSeaGrass** [TanStack/query#5188] [Q&A]
I'm trying to generically type my API call so that I can reuse, however the mutation code (which was previously fine, before typing) is now  kicking up an error:

No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ mutationFn: <T extends MethodName>({ payload, method, }: { payload: any; method: T; }) => Promise<ResponseType<T> | undefined>; }' is not assignable to parameter of type 'MutationKey'.
      Object literal may only specify known properties, and 'mutationFn' does not exist in type 'readonly unknown[]'.

I've tried adding a mutation key before mutationFn but no luck. I've also seen that multiple params seem to produce this error however I believe I'm passing in an object.

Any help appreciated :)

Edit: clicking the link doesn't work but pasting does! (please ignore rependency error as should illustrate error)

I've created a simplified version of the code here: [https://codesandbox.io/embed/typescript-playground-export-forked-yvmu69?fontsize=14&hidenavigation=1&theme=dark](https://codesandbox.io/embed/typescript-playground-export-forked-yvmu69?fontsize=14&hidenavigation=1&theme=dark)

**Comment by ThreeSeaGrass** [comment]
Edited link

**Comment by TkDodo** [comment]
Hm, I don't really understand the problem, but I've narrowed it down to an even simpler example. It has nothing to do with the conditional return type, but with the generic on the input:

https://www.typescriptlang.org/play#code/JYWwDg9gTgLgBAbzgVwM4FMCyyYEMbAQB2cAvnAGZQQhwBEAAnkangMYDWA9FOrmzAC0AR2TooATzoBYAFByYEsOjiZ0MABYQAJgDlcIFQF56hzToCMdOAB9T6rdoBMdANxy5bYqzgx0rADFkIjYLOBNcVAkQuAAeABU4dAAPPyJtVFUHHX1DAD4ACmAiMBwALjh4gEoKgAVqEGAMWNwiCTzwjoQ5ODheGGQoEiJkABtRuEjJtrlSd3lZLxZ4btlekBx8dABBKJCKtl4t2twJQyIYAEkL9AvsPAJiABoeuCaAGQhcbWKAcwOjn5rmkYJ9vn9ZuEUBh7vhCEQCqt1ptHkQAkQKn5AsFQrMqvMgA

It works if we remove `T extends MethodName` ...

**Reply by hgmeza** [reply]
Hi @TkDodo,

I am trying to also the same thing of using a generic type so that it returns a type based on some conditions, similar to the example you have posted above. Since we still have some errors in that example
```
Type 'void' is not assignable to type 'MethodName'
```
would you say that `useMutation` does not yet support generic inputs on the `mutationFn`? I am having exactly that same error.

Would love your input on this.

**Reply by mrlubos** [reply]
@TkDodo I have the same problem. My mutation function can return different responses based on the provided field. For example, `foo('a')` might return `{ a: 1 }`, `foo('b')` might return `{ b: 1 }`, etc. Are you able to suggest how to resolve this problem?
